lib/rucio/daemons/storage/consistency/actions.py

```diff
@@ -0,0 +1,805 @@
+        if os.stat(path).st_size > 0:
+            r, timestamp, typ, ext = parse_filename(fn)
+            if r == rse:
+                # if the RSE was X, then rses like X_Y will appear in this list too,
+                # so double check that we get the right RSE
+                runs.append(path)
+    if nlast == 0:
+        nlast = len(runs)
+    return sorted(runs, reverse=False)[-nlast:]
+
+def list_unprocessed_runs(path, rse, nlast=0):
+    files = glob.glob(f"{path}/{rse}_*_stats.json")
+    unproc_runs = []
+    for path in files:
+        fn = path.rsplit("/", 1)[-1]
+        if os.stat(path).st_size > 0:
+            r, timestamp, typ, ext = parse_filename(fn)
+            if r == rse:
+                # if the RSE was X, then rses like X_Y will appear in this list too,
+                # so double check that we get the right RSE
+                if not was_cc_attempted(path):
+                    unproc_runs.append(timestamp)
+    if nlast == 0:
+        nlast = len(unproc_runs)
+    return sorted(unproc_runs, reverse=True)[-nlast:]
+
+def was_cc_attempted(stats_file):
+    try:
+        f = open(stats_file, "r")
+    except:
+        print("get_data: error ", stats_file)
+        return None
+    stats = json.loads(f.read())
+    cc_dark_status = ''
+    cc_miss_status = ''
+    if "cc_dark" in stats or "cc_miss" in stats:
+        return True
+    else:
+        return False
+
+def was_cc_processed(stats_file):
+    try:
+        f = open(stats_file, "r")
+    except:
+        print("get_data: error ", stats_file)
+        return None
+    stats = json.loads(f.read())
+    cc_dark_status = ''
+    cc_miss_status = ''
+    if "cc_dark" in stats:
+        if "status" in stats['cc_dark']:
+            cc_dark_status = stats['cc_dark']['status']
+    if "cc_miss" in stats:
+        if "status" in stats['cc_miss']:
+            cc_miss_status = stats['cc_miss']['status']
+    if cc_dark_status == 'done' or cc_miss_status == 'done':
+        return True
+    else:
+        return False
+
+
+def process_dark_files(path, rse, latest_run, maxdarkfraction, max_files_at_site, oldenough_run):
+
+    """
+    Process the Dark Files.
+    """
+
+    prefix = 'storage-consistency-actions (process_dark_files())'
+    logger = formatted_logger(logging.log, prefix + '%s')
+
+# Create a cc_dark section in the stats file
+
+    t0 = time.time()
+    stats_key = "cc_dark"
+    cc_stats = stats = None
+    stats = Stats(latest_run)
+    cc_stats = {
+        "start_time": t0,
+        "end_time": None,
+        "initial_dark_files": 0,
+        "confirmed_dark_files": 0,
+        "x-check_run": oldenough_run,
+        "status": "started"
+    }
+    stats[stats_key] = cc_stats
+
+# Compare the two lists, and take only the dark files that are in both
+    latest_dark = re.sub('_stats.json$', '_D.list', latest_run)
+    oldenough_dark = re.sub('_stats.json$', '_D.list', oldenough_run)
+    logger(logging.INFO,'\n latest_dark = %s' % latest_dark)
```

Comment:
Are the new line characters needed? This makes the log output harder to read.

Reply:
I can change that. I was using it consistently in my code as the log is MORE readable for me that way. But if that makes it more difficult for everybody else, I'll change it.

Reply:
Done.